p2p/crypto/secio/rw.go

```diff
@@ -105,66 +101,27 @@ func (r *etmReader) NextMsgLen() (int, error) {
 	return r.msg.NextMsgLen()
 }
 
-func (r *etmReader) drainBuf(buf []byte) int {
-	if r.buf == nil {
-		return 0
-	}
-
-	n := copy(buf, r.buf)
-	r.buf = r.buf[n:]
-	return n
-}
-
 func (r *etmReader) Read(buf []byte) (int, error) {
 	r.Lock()
 	defer r.Unlock()
 
-	// first, check if we have anything in the buffer
-	copied := r.drainBuf(buf)
-	buf = buf[copied:]
```

Comment:
Ah, thanks for the explanation. This is me being a go newbie, despite being an experience C programmer. I was thinking that somehow `buf` was manipulating a reference and `buf = buf[copied:]` was changing the result that the caller sees.